File: src/cpp/stochastic_models.cpp
#include "stochastic_models.h"
#include <cmath>
#include <random>

std::vector<double> geometricBrownianMotion(double S0, double mu, double sigma, double T, int numSteps) {
    std::vector<double> prices(numSteps + 1, S0);
    double dt = T / numSteps;
    std::random_device rd;
    std::mt19937 gen(rd());
    std::normal_distribution<> dist(0.0, 1.0);

    for (int i = 1; i <= numSteps; ++i) {
        double randomShock = dist(gen);
        prices[i] = prices[i - 1] * std::exp((mu - 0.5 * sigma * sigma) * dt + sigma * std::sqrt(dt) * randomShock);
    }

    return prices;
}

File: src/python/stochastic_models.py
import numpy as np

def geometric_brownian_motion(S0, mu, sigma, T, num_steps):
    """
    Simulates Geometric Brownian Motion for asset prices.
    """
    dt = T / num_steps
    prices = np.zeros(num_steps)
    prices[0] = S0

    for t in range(1, num_steps):
        random_shock = np.random.normal(0, np.sqrt(dt))
        prices[t] = prices[t - 1] * np.exp((mu - 0.5 * sigma**2) * dt + sigma * random_shock)
    
    return prices

def mean_reverting_process(S0, theta, mu, sigma, T, num_steps):
    """
    Simulates a mean-reverting process (e.g., Ornstein-Uhlenbeck).
    """
    dt = T / num_steps
    prices = np.zeros(num_steps)
    prices[0] = S0

    for t in range(1, num_steps):
        random_shock = np.random.normal(0, np.sqrt(dt))
        prices[t] = prices[t - 1] + theta * (mu - prices[t - 1]) * dt + sigma * random_shock
    
    return prices

